qutip/bloch.py

```diff
@@ -56,7 +60,12 @@ def __init__(self, xs, ys, zs, *args, **kwargs):
 
         def draw(self, renderer):
             xs3d, ys3d, zs3d = self._verts3d
-            xs, ys, zs = proj3d.proj_transform(xs3d, ys3d, zs3d, renderer.M)
+            if parse_version(matplotlib.__version__) >= parse_version('3.4'):
```

Comment:
`guide-bloch.rst` doesn't need a defended import of `matplotlib` because that's a dependency of the documentation build - it'll always be there.

The alternative is to keep the test like this here, but in all the places where it's about adding the `auto_add_to_figure=False` kwarg to `Axes3D`, you could do the test once at the top of the file, and rebind
```python
if [ ... matplotlib >= 3.4 ... ]:
    def Axes3D(*args, **kwargs):
        return mpl_toolkits.mplot3d.Axes3D(*args, auto_add_to_figure=False, **kwargs)
else:
    def Axes3D(*args, **kwargs):
        return mpl_toolkits.mplot3d.Axes3D(*args, **kwargs)
```
or something like that.

Reply:
Also, I didn't test, but what's the earliest version of `matplotlib` that `axes.M` works in?  If it works in mpl 3.0, then we can ellide the version test in this particular instance entirely, and just use `axes.M` in all cases.

Reply:
Thank you so much for the review @jakelishman ! In the recent commit, I have incorporated the reviously requested changes. I will surely comprehend your suggestions.

> The alternative is to keep the test like this here, but in all the places where it's about adding the auto_add_to_figure=False kwarg to Axes3D, you could do the test once at the top of the file, and rebind

Thank you for this suggestion! I was wondering whether this could take more time for it's instantiation since then we would check versions for every call to `Axes3D`, given that the places where such conditions are required are sparsely spread across the module? I might be missing on some places, and hence would love to get your feedback!

Reply:
> Also, I didn't test, but what's the earliest version of matplotlib that axes.M works in? If it works in mpl 3.0, then we can ellide the version test in this particular instance entirely, and just use axes.M in all cases.

@jakelishman I checked with `matplotlib==3.0.1` and `self.axes.M` does work. Using `matplotlib==3.0.0` with `python>3.6.6` gives some import errors but was fixed in `matplotlib==3.0.1` (as mentioned [here](https://stackoverflow.com/questions/52979322/matplotlib-3-0-0-cannot-import-name-get-backend-from-matplotlib)).

In any case, you had said that if it works for mpl 3.0, then we could omit version tests for `proj3d.proj_transform` which uses `axes.M`. Could you let me know the reason behind this?

Thanks!

Reply:
> `guide-bloch.rst` doesn't need a defended import of `matplotlib` because that's a dependency of the documentation build - it'll always be there. (_edit_: just realised this wasn't about the defended import at all, but the versioning. I'm actually happy for the documentation to rely on matplotlib 3.4 if it makes things clearer; building the documentation isn't the same as running in a production environment, and I'd rather the documentation be legible without additional version checks. That said, we'll have to update the `requirements.txt` file, and that's a shade non-trivial, because there's actually some pinned versions in there that are very important.)
> 
> The alternative is to keep the test like this here, but in all the places where it's about adding the `auto_add_to_figure=False` kwarg to `Axes3D`, you could do the test once at the top of the file, and rebind
> 
> ```python
> if [ ... matplotlib >= 3.4 ... ]:
>     def Axes3D(*args, **kwargs):
>         return mpl_toolkits.mplot3d.Axes3D(*args, auto_add_to_figure=False, **kwargs)
> else:
>     def Axes3D(*args, **kwargs):
>         return mpl_toolkits.mplot3d.Axes3D(*args, **kwargs)
> ```
> 
> or something like that.


The above suggested change by @jakelishman seems to provide better readability than the current implementation in this PR, especially because the internal implementation of the function do not change within the module. I would make those changes and analyse its outcomes. Thanks!

Reply:
> Thank you for this suggestion! I was wondering whether this could take more time for it's instantiation since then we would check versions for every call to `Axes3D`

The defended `def Axes3D` statement I did only tests the version once, at module initialisation, and then saves it for use later.  You don't need to worry about performance concerns between a single `if` statement and wrapper Python function in general, so prefer the one that's more readable (as you say).

> I checked with `matplotlib==3.0.1` and `self.axes.M` does work. Using `matplotlib==3.0.0` with `python>3.6.6` gives some import errors but was fixed in `matplotlib==3.0.1`

In that case, definitely better to prefer just using `self.axes.M` in all cases.  Don't worry if matplotlib messed up their 3.0.0 release - we still support it, it's just that 3.0.0 happens to be broken with new versions of Python.

> In any case, you had said that if it works for mpl 3.0, then we could omit version tests for `proj3d.proj_transform` which uses `axes.M`. Could you let me know the reason behind this?

In an ideal world, there would only be one way to do things in a library, and we would never have to defensively programme against different versions.  So, if there's a recommended way that works with all our supported versions, we prefer to do that and have as few version checks as we can - every time we have a version check, it means we've got duplicated code, which is harder to read and harder to maintain.

Reply:
Oh, I should also note that my original redefinition of `Axes3D` should probably have included a catch for the figure and called `figure.add_axes(axes)` call inside it.